pkg/scalers/splunk_o11y_scaler.go

```diff
@@ -0,0 +1,264 @@
+		meta.realm = realm
+	} else {
+		return nil, fmt.Errorf("no realm given")
+	}
+	logger.Info("Splunk Realm -> %s\n", meta.realm)
+
+	// Debug TODO check
+	meta.vType = v2.ValueMetricType
+
+	return &meta, nil
+}
+
+func newSplunkO11yConnection(ctx context.Context, meta *splunkO11yMetadata, config *scalersconfig.ScalerConfig) (*signalflow.Client, error) {
+	accessToken := meta.accessToken
+	realm := meta.realm
+
+	if realm == "" || accessToken == "" {
+		return nil, fmt.Errorf("error. could not find splunk access token or ream.")
+	}
+
+	apiClient, err := signalflow.NewClient(
+		signalflow.StreamURLForRealm(realm),
+		signalflow.AccessToken(accessToken))
+	if err != nil {
+		return nil, fmt.Errorf("error creating SignalFlow client: %w", err)
+	}
+
+	return apiClient, nil
+}
+
+func logMessage(logger logr.Logger, msg string, value float64) {
+	if value != -1 {
+		msg = fmt.Sprintf("splunk_o11y_scaler: %s -> %v", msg, value)
+	} else {
+		msg = fmt.Sprintf("splunk_o11y_scaler: %s", msg)
+	}
+	logger.Info(msg)
+}
+
+func (s *splunkO11yScaler) getQueryResult(ctx context.Context) (float64, error) {
+	// var duration time.Duration = 1000000000 // one second in nano seconds
+	var duration time.Duration = 10000000000 // ten seconds in nano seconds
+
+	comp, err := s.apiClient.Execute(context.Background(), &signalflow.ExecuteRequest{
+		Program: s.metadata.query,
+	})
+	if err != nil {
+		return -1, fmt.Errorf("error: could not execute signalflow query: %w", err)
+	}
+
+	go func() {
+		time.Sleep(duration)
+		if err := comp.Stop(context.Background()); err != nil {
+			s.logger.Info("Failed to stop computation")
+		}
+	}()
+
+	logMessage(s.logger, "Received Splunk Observability metrics", -1)
+
+	max := math.Inf(-1)
+	min := math.Inf(1)
+	valueSum := 0.0
+	valueCount := 0
+	s.logger.Info("getQueryResult -> Now Iterating")
+	for msg := range comp.Data() {
+		s.logger.Info("getQueryResult -> msg: %+v\n", msg)
+		if len(msg.Payloads) == 0 {
+			s.logger.Info("getQueryResult -> No data retreived. Continuing")
+			continue
+		}
+		for _, pl := range msg.Payloads {
+			value, ok := pl.Value().(float64)
+			if !ok {
+				return -1, fmt.Errorf("error: could not convert Splunk Observability metric value to float64")
+			}
+			logMessage(s.logger, "Encountering value ", value)
+			if value > max {
+				max = value
+			}
+			if value < min {
+				min = value
+			}
+			valueSum += value
+			valueCount++
+		}
+	}
+
+	if valueCount > 1 && s.metadata.queryAggegrator == "" {
+		return 0, fmt.Errorf("query returned more than 1 series; modify the query to return only 1 series or add a queryAggregator")
+	}
+
+	switch s.metadata.queryAggegrator {
+	case "max":
+		logMessage(s.logger, "Returning max value ", max)
+		return max, nil
+	case "min":
+		logMessage(s.logger, "Returning min value ", min)
+		return min, nil
+	case "avg":
+		avg := valueSum / float64(valueCount)
+		logMessage(s.logger, "Returning avg value ", avg)
+		return avg, nil
+	default:
+		return max, nil
```

Comment:
Can these aggregations be performed as part of the Splunk query? I think it would be more reliable to use server side logic.

Reply:
I will look into that.